core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -700,7 +702,7 @@ sealed trait ZIO[-R, +E, +A]
    * failure except external interruption.
    */
   final def foldCause[B](failure: Cause[E] => B, success: A => B)(implicit trace: Trace): URIO[R, B] =
-    foldCauseZIO(c => Exit.succeed(failure(c)), a => Exit.succeed(success(a)))
+    foldCauseZIO(c => Exit.succeed(c.failureOrCause.fold(???, failure(_))), a => Exit.succeed(success(a)))
```

Comment:
The problem described in this issue is not with `foldCause`. From what I can tell it has to do with `Cause#toString` when there is a large number of Causes that were combined together.

Reply:
@kyri-petrou I am curious if you are refering to somewhere here? https://github.com/EstebanMarin/zio/blob/3403a221821863b3babd897c1ad26e4c2a70e998/core/shared/src/main/scala/zio/Cause.scala#L713

Reply:
I am unsure of the place, I am looking at the Cause class to see any toString method, but I only got this.

Reply:
The problem is that the `toString` method of `Cause` (in this case, `Cause.Both`) is called recursively, causing an SO error. We need to override the `toString` method and make it stack-safe